File: starters/kotlin-android/app/src/main/java/org/example/foundation/kotlin/HttpTaskRepository.kt
package org.example.foundation.kotlin

import org.example.foundation.core.application.TaskRepository
import org.example.foundation.core.domain.Task
import org.example.foundation.core.domain.TaskError
import org.example.foundation.core.domain.TaskResult
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URI

/** Android transport adapter. The core has no JSON, HTTP, SQL or framework dependency. */
class HttpTaskRepository(baseUrl: String, private val tokenProvider: () -> String? = { null }) : TaskRepository {
    private val base = URI(baseUrl)
    private var snapshots = mutableMapOf<String, Pair<Task, Int>>()
    private var bound = false
    private var sessionToken: String? = null
    init {
        require(base.scheme == "https" || (base.scheme == "http" && base.host in listOf("127.0.0.1", "localhost", "[::1]", "10.0.2.2")))
        require(base.userInfo == null && base.query == null && base.fragment == null && base.path.matches(Regex(".*/api/v1/?")))
    }

    private fun request(path: String, method: String = "GET", body: JSONObject? = null): JSONObject {
        val connection = URI(base.toString().trimEnd('/') + path).toURL().openConnection() as HttpURLConnection
        try {
            connection.requestMethod = method
            connection.instanceFollowRedirects = false
            connection.connectTimeout = 10000
            connection.readTimeout = 10000
            connection.useCaches = false
            connection.setRequestProperty("Accept", "application/json")
            val currentToken = tokenProvider()
            if (bound && currentToken != sessionToken) {
                snapshots.clear()
                error("Session changed; recreate the repository and clear UI state.")
            }
            bound = true
            sessionToken = currentToken
            currentToken?.let { token ->
                require(token.length == 64 && token.matches(Regex("[0-9a-f]{64}")))
                connection.setRequestProperty("Authorization", "Bearer $token")
            }
            if (body != null) {
                connection.doOutput = true
                connection.setRequestProperty("Content-Type", "application/json")
                val bytes = body.toString().toByteArray(Charsets.UTF_8)
                connection.setFixedLengthStreamingMode(bytes.size)
                connection.outputStream.use { it.write(bytes) }
            }
            check(connection.responseCode == if (method == "POST") 201 else 200)
            check(connection.contentType?.startsWith("application/json") == true)
            val bytes = connection.inputStream.use { input ->
                val output = java.io.ByteArrayOutputStream()
                val buffer = ByteArray(8192)
                while (true) {
                    val count = input.read(buffer)
                    if (count == -1) break
                    check(output.size() + count <= 1048576)
                    output.write(buffer, 0, count)
                }
                output.toByteArray()
            }
            val text = Charsets.UTF_8.newDecoder().decode(java.nio.ByteBuffer.wrap(bytes)).toString()
            return JSONObject(text)
        } finally { connection.disconnect() }
    }

    private fun decode(row: JSONObject): Pair<Task, Int> {
        val id = row.get("id")
        val title = row.get("title")
        val completed = row.get("completed")
        val version = row.get("version")
        require(id is String && id.matches(Regex("[0-9a-f]{8}(-[0-9a-f]{4}){3}-[0-9a-f]{12}")))
        require(title is String && completed is Boolean && version is Int && version in 1..2147483646)
        return Task(id, title, completed, null) to version
    }

    private inline fun <T> run(operation: () -> T): TaskResult<T> = try {
        TaskResult.Success(operation())
    } catch (_: Exception) {
        // No automatic retry, including conflicts and uncertain committed writes.
        TaskResult.Failure(TaskError.STORAGE_FAILURE)
    }

    override fun list(): TaskResult<List<Task>> = run {
        val rows = linkedMapOf<String, Pair<Task, Int>>()
        var after: String? = null
        var finished = false
        for (page in 0 until 100) {
            val response = request("/tasks?limit=100" + (after?.let { "&after=$it" } ?: ""))
            val data = response.getJSONArray("data")
            require(data.length() <= 100)
            if (data.length() == 0) {
                require(response.has("next_after") && response.isNull("next_after"))
                finished = true
                break
            }
            for (index in 0 until data.length()) {
                val row = decode(data.getJSONObject(index))
                require(!rows.containsKey(row.first.id) && (after == null || row.first.id > after))
                rows[row.first.id] = row
                after = row.first.id
            }
            require(response.get("next_after") == after)
        }
        check(finished)
        snapshots = rows
        rows.values.map { it.first }
    }

    override fun add(task: Task): TaskResult<Task> = run {
        val row = decode(request("/tasks", "POST", JSONObject().put("title", task.title)).getJSONObject("data"))
        snapshots[row.first.id] = row
        row.first
    }

    override fun update(id: String, transform: (Task) -> Task): TaskResult<Task> = run {
        val current = checkNotNull(snapshots[id])
        val next = transform(current.first)
        require(next.id == id && next.title == current.first.title && current.second < 2147483646)
        val row = decode(request("/tasks/$id", "PUT", JSONObject()
            .put("title", next.title).put("completed", next.completed).put("version", current.second)).getJSONObject("data"))
        check(row.first.id == id && row.first.title == next.title && row.first.completed == next.completed && row.second == current.second + 1)
        snapshots[id] = row
        row.first
    }
}

File: starters/flutter/lib/infrastructure/http_task_repository.dart
import 'dart:convert';
import 'dart:io';
import '../application/task_repository.dart';
import '../domain/task.dart';

/// HTTP is an adapter only; no server/framework or database types cross the port.
final class HttpTaskRepository implements TaskRepository {
  HttpTaskRepository(String url, {String? Function()? tokenProvider})
    : _tokenProvider = tokenProvider ?? (() => null),
      base = Uri.parse(url) {
    if ((base.scheme != 'https' &&
            !(base.scheme == 'http' &&
                [
                  '127.0.0.1',
                  'localhost',
                  '::1',
                  '10.0.2.2',
                ].contains(base.host))) ||
        base.userInfo.isNotEmpty ||
        base.hasQuery ||
        base.hasFragment ||
        !RegExp(r'/api/v1/?$').hasMatch(base.path)) {
      throw ArgumentError(
        'Use an explicit HTTPS /api/v1 URL or local development loopback.',
      );
    }
  }
  final Uri base;
  final String? Function() _tokenProvider;
  bool _bound = false;
  String? _sessionToken;
  Map<String, Map<String, dynamic>> _snapshots = {};
  static final _id = RegExp(r'^[0-9a-f]{8}(-[0-9a-f]{4}){3}-[0-9a-f]{12}$');

  Map<String, dynamic> _decode(dynamic value) {
    if (value is! Map<String, dynamic> ||
        value['id'] is! String ||
        (value['id'] as String).length != 36 ||
        !_id.hasMatch(value['id']) ||
        value['title'] is! String ||
        value['completed'] is! bool ||
        value['version'] is! int ||
        value['version'] < 1 ||
        value['version'] > 2147483646) {
      throw const TaskException(TaskIssue.unavailable);
    }
    final title = TaskTitle(value['title']);
    if (title.value != value['title']) {
      throw const TaskException(TaskIssue.unavailable);
    }
    return Map.unmodifiable(value);
  }

  TaskItem _item(Map<String, dynamic> row) => TaskItem(
    id: row['id'],
    title: TaskTitle(row['title']),
    completed: row['completed'],
  );

  Future<dynamic> _request(
    String path,
    String method, [
    Map<String, dynamic>? body,
  ]) async {
    final client = HttpClient()
      ..connectionTimeout = const Duration(seconds: 10);
    try {
      return await (() async {
        final url = Uri.parse(
          '${base.toString().replaceFirst(RegExp(r'/$'), '')}$path',
        );
        final request = await client.openUrl(method, url);
        final token = _tokenProvider();
        if (_bound && token != _sessionToken) {
          _snapshots = {};
          throw const TaskException(TaskIssue.unavailable);
        }
        _bound = true;
        _sessionToken = token;
        if (token != null) {
          if (token.length != 64 ||
              !RegExp(r'^[0-9a-f]{64}$').hasMatch(token)) {
            throw const TaskException(TaskIssue.unavailable);
          }
          request.headers.set(HttpHeaders.authorizationHeader, 'Bearer $token');
        }
        request.followRedirects = false;
        request.headers.set(HttpHeaders.acceptHeader, 'application/json');
        if (body != null) {
          request.headers.contentType = ContentType.json;
          request.write(jsonEncode(body));
        }
        final response = await request.close();
        if (response.statusCode == 404) {
          throw const TaskException(TaskIssue.notFound);
        }
        if (response.statusCode != (method == 'POST' ? 201 : 200) ||
            response.headers.contentType?.mimeType != 'application/json') {
          throw const TaskException(TaskIssue.unavailable);
        }
        final bytes = <int>[];
        await for (final chunk in response) {
          bytes.addAll(chunk);
          if (bytes.length > 1048576) {
            throw const TaskException(TaskIssue.unavailable);
          }
        }
        return jsonDecode(utf8.decode(bytes));
      })().timeout(const Duration(seconds: 10));
    } on TaskException {
      rethrow;
    } catch (_) {
      throw const TaskException(TaskIssue.unavailable);
    } finally {
      client.close(force: true);
    }
  }

  @override
  Future<List<TaskItem>> readAll() async {
    final rows = <String, Map<String, dynamic>>{};
    String? after;
    for (var page = 0; page < 100; page++) {
      final response = await _request(
        '/tasks?limit=100${after == null ? '' : '&after=$after'}',
        'GET',
      );
      if (response is! Map ||
          response['data'] is! List ||
          (response['data'] as List).length > 100) {
        throw const TaskException(TaskIssue.unavailable);
      }
      for (final value in response['data']) {
        final row = _decode(value);
        final id = row['id'] as String;
        if (rows.containsKey(id) ||
            (after != null && id.compareTo(after) <= 0)) {
          throw const TaskException(TaskIssue.unavailable);
        }
        rows[id] = row;
        after = id;
      }
      if ((response['data'] as List).isEmpty) {
        if (!response.containsKey('next_after') ||
            response['next_after'] != null) {
          throw const TaskException(TaskIssue.unavailable);
        }
        _snapshots = rows;
        return List.unmodifiable(rows.values.map(_item));
      }
      if (response['next_after'] != after) {
        throw const TaskException(TaskIssue.unavailable);
      }
    }
    throw const TaskException(TaskIssue.unavailable);
  }

  @override
  Future<TaskItem> add(TaskTitle title) async {
    final row = _decode(
      (await _request('/tasks', 'POST', {'title': title.value}))['data'],
    );
    _snapshots[row['id']] = row;
    return _item(row);
  }

  @override
  Future<TaskItem> setCompleted(String id, bool completed) async {
    final current = _snapshots[id];
    if (current == null) throw const TaskException(TaskIssue.notFound);
    if (current['version'] >= 2147483646) {
      throw const TaskException(TaskIssue.unavailable);
    }
    final row = _decode(
      (await _request('/tasks/$id', 'PUT', {
        'title': current['title'],
        'completed': completed,
        'version': current['version'],
      }))['data'],
    );
    if (row['id'] != id ||
        row['version'] != current['version'] + 1 ||
        row['title'] != current['title'] ||
        row['completed'] != completed) {
      throw const TaskException(TaskIssue.unavailable);
    }
    _snapshots[id] = row;
    return _item(row);
  }
}
